rfd/0194-k8s-rbac-simplification.md

```diff
@@ -0,0 +1,210 @@
+  - <https://github.com/gravitational/teleport/issues/45475>
+  - (internal) <https://goteleport.slack.com/archives/C0582MBSMHN/p1738187497666819>
+- Unexpected `exec` grant on read-only user:
+  (internal) <https://gravitational.slack.com/archives/C32M8FP1V/p1739462454321459?thread_ts=1739462384.714419&cid=C32M8FP1V>
+
+## Goals
+
+- Make the Teleport agent authoritative
+  - Handle permissions for the users instead of relying on impersonating a
+    different user/group
+- Discourage (or even deprecate?) the use of the `kubernetes_groups` and
+  `kubernetes_users` in favor of `kubernetes_resources`
+- Clarify / Improve documentation for the various ways to setup Kubernetes in
+  Teleport.
+
+## Proposal
+
+### Background
+
+Currently, when enrolling a Kubernetes cluster and setting it up to use the
+group `system:masters`, everything works as expected. Teleport is super admin
+and yields reduced permissions to the user based on the `kubernetes_resources`
+and `kubernetes_label` fields from the Teleport role.
+
+### Impersonation
+
+To limit the number of code changes, the proposal is to default the
+impersonation to use the Teleport service account.
+
+We will need to make sure the service account has `cluster-admin` permissions.
+
+To avoid unexpected access being granted, the `kubernetes_resources` field
+should default to `null` as well instead of the current wildcard.
+
+The `kubernetes_user` should default to the Teleport user auth email.
+
+### UX
+
+#### Deprecate `namespace`
+
+We currently have the `namespace` kind mean "everything within a namespace",
+which has caused confusion with customers and resulting in unexpected access
+granted.
+
+This should be deprecated and removed. It would require a bump on the role
+model to v8 to avoid changing behavior for current customers.
+
+Upgrading to v8 would map the deprecated `namespace` field to `*` while
+swapping `name`/`namespace` as these are equivalent within a namespace:
+
+```yaml
+kind: namespace
+name: foo
+```
+
+and:
+
+```yaml
+kind: '*'
+namespace: foo
+```
+
+Note that the later also grants access to cluster-wide resources outside a
+namesace.
+
+#### CRD / Subresources support
+
+Currently, we only support a hard-coded list of pre-set resources. To enable
+CRD support, i.e. arbitrary kind, a new field `api` will be introduced.
+When not set, the `kind` field would still be restricted to the existing
+hard-coded list, but once set, it can be an arbitrary string.
+
+Similirally, to avoid confusion around sub-resources and the common mistake
+of granting `exec` permission with the `get` verb, a new field `sub_resources`
+will be introduce. If not set, then only the main resource would match, which
+removes the unexpected side-effect for `exec`.
+To avoid confusion, when `sub_resources` are set, the `verbs` field will be
+ignored.
+
+The role would look like this:
+
+```yaml
+kubernetes_resources:
+  - kind: '*'
+    api: 'API with version'
```

Comment:
The version is not relevant in this context. This occurs because Kubernetes uses version webhooks to automatically convert resources from the storage version to the requested version.  

For example, if `api/v2` is blocked, a user who requests `api/v1` instead will still receive the v1 representation of the object. Kubernetes RBAC is version agnostic similar to how teleport should be

Reply:
hmm, so we shouldn't support blocking api/v2 while still allowing api/v1?

Reply:
no. The underlying object is the same so we should block access to v1 and v2